chainio/txmgr/geometric/geometric.go

```diff
@@ -0,0 +1,598 @@
+// It does gas estimation and gas price bumping to ensure the transaction gets mined,
+// but it does not do nonce management, so the tx argument must have the correct nonce already set.
+//
+// Send is blocking and safe to call concurrently, so sending multiple txs in parallel is safe.
+func (t *GeometricTxManager) Send(ctx context.Context, tx *types.Transaction) (*types.Receipt, error) {
+	return t.processTransaction(ctx, newTxnRequest(tx))
+}
+
+// ProcessTransaction sends the transaction and runs the monitoring loop which will bump the gasPrice until the tx get
+// included.
+func (t *GeometricTxManager) processTransaction(ctx context.Context, req *txnRequest) (*types.Receipt, error) {
+	t.logger.Debug("new transaction",
+		"nonce", req.tx.Nonce(), "gasFeeCap", req.tx.GasFeeCap(), "gasTipCap", req.tx.GasTipCap(),
+	)
+	t.metrics.IncrementProcessingTxCount()
+	defer t.metrics.DecrementProcessingTxCount()
+
+	var txn *types.Transaction
+	var txID wallet.TxID
+	var err error
+	retryFromFailure := 0
+	from, err := t.wallet.SenderAddress(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get sender address: %w", err)
+	}
+	for retryFromFailure < t.params.MaxSendTransactionRetry {
+		gasTipCap, err := t.estimateGasTipCap(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to estimate gas tip cap: %w", err)
+		}
+		txn, err = t.updateGasTipCap(ctx, req.tx, gasTipCap, from)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update gas price: %w", err)
+		}
+		txID, err = t.wallet.SendTransaction(ctx, txn)
+		// TODO: what is this...? does it come from the fireblocks wallet?
+		var timeout interface{ Timeout() bool }
+		if errors.As(err, &timeout); timeout != nil && timeout.Timeout() {
+			t.logger.Warn(
+				"failed to send txn due to timeout",
+				"hash",
+				txn.Hash().Hex(),
+				"numRetries",
+				retryFromFailure,
+				"maxRetry",
+				t.params.MaxSendTransactionRetry,
+				"err",
+				err,
+			)
+			// TODO: why do we only retry on client or server timeouts? what about other errors?
+			retryFromFailure++
+			continue
+		} else if err != nil {
+			return nil, fmt.Errorf("failed to send txn %s: %w", txn.Hash().Hex(), err)
+		} else {
+			t.logger.Debug("successfully sent txn", "txID", txID, "txHash", txn.Hash().Hex())
+			break
+		}
+	}
+
+	if txn == nil || txID == "" {
+		return nil, fmt.Errorf("failed to send txn %s: %w", req.tx.Hash().Hex(), err)
+	}
+
+	req.tx = txn
+	req.txAttempts = append(req.txAttempts, &transaction{
+		TxID:        txID,
+		Transaction: txn,
+		requestedAt: time.Now(),
+	})
+
+	receipt, err := t.monitorTransaction(ctx, req)
+	if err == nil {
+		if receipt.GasUsed > 0 {
+			t.metrics.ObserveGasUsedWei(receipt.GasUsed)
+		}
+		t.metrics.ObserveConfirmationLatencyMs(time.Since(req.requestedAt).Milliseconds())
+	}
+	return receipt, err
+}
+
+// ensureAnyTransactionBroadcasted waits until all given transactions are broadcasted to the network.
```

Comment:
`ensureAnyTransactionBroadcasted` 
waits until all given transactions are broadcasted to the network

is this comment correct? methods says `any` but description says `all` given txs?

Reply:
Good question. @ian-shim I'm also not sure how this function works. It seems to return nil as soon as one function returns an error, which sounds like a bug?
```
		for _, tx := range txs {
			_, err := t.wallet.GetTransactionReceipt(ctx, tx.TxID)
			if err == nil || errors.Is(err, ethereum.NotFound) || errors.Is(err, wallet.ErrReceiptNotYetAvailable) {
				t.metrics.ObserveBroadcastLatencyMs(time.Since(tx.requestedAt).Milliseconds())
				return nil
			}
		}
```
Basically if any one of the txs hasn't been mined (ethereum.NotFound is returned) then we return. But this could also happen if we have a bug and are querying a random (wrong) txHash, so it feels wrong?

This function seems like it should only be called when the wallet is fireblocks? Should we change the code to reflect that?

Reply:
The comment seems wrong. It returns as long as any txn has been broadcasted. 

> This function seems like it should only be called when the wallet is fireblocks?

Correct!

Reply:
Ok thanks! I fixed the comment [e929bf5](https://github.com/Layr-Labs/eigensdk-go/pull/284/commits/e929bf5feb7d21929b5ede2cb190bc9907d2b83d)
But I'm still a bit confused @ian-shim . waiting for broadcast is only useful for fireblocks, why are we only checking that its a fireblocksWallet [here](https://github.com/Layr-Labs/eigensdk-go/blob/samlaf-avs-588-upstream-tx-manager-to-eigensdk-go/chainio/txmgr/geometric/geometric.go#L375)? Instead of above. Aka why ae we calling ensureAnyTxBroadcasted and doing the whole huge if statement below for other wallets?

But also overall I'm not convinced this function is actually doing something useful. Maybe I'm missing somethingl... but [here](https://github.com/Layr-Labs/eigensdk-go/blob/samlaf-avs-588-upstream-tx-manager-to-eigensdk-go/chainio/txmgr/geometric/geometric.go#L258) we will just return eagerly if any tx has not been mined yet. But not being mined doesn't mean its been broadcasted to the network and is in the mempool right? What we would want is to check if a tx is in the mempool. I think `eth_getTransactionCount` factors in the txs in the mempool, so we could check that the txCount returned is >= the nonce of the tx? Am I missing something here?

Reply:
@ian-shim after our discussion this is what I came up with: [072d670](https://github.com/Layr-Labs/eigensdk-go/pull/284/commits/072d670dfeb786dbe5cb9ea9bc2e6fe01d20505a)
Can you quickly check that this makes sense?

Reply:
Thanks, that makes sense!